### PR TITLE
http2: various updates and improvements

### PR DESCRIPTION
**Pull request by jasnell**
* Lint fixes
* Improving http/2 compliance
* Improving efficiency (do not send unnecessary empty data frames, etc)

(more coming)

##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by sebdeckers** [comment]
Oops didn't see this... just did the same linting cleanup in #163 (amongst other things).

**Comment by sebdeckers** [comment]
@jasnell btw maybe take a look at this compiler warning:

```
In file included from ../src/node_http2_core.cc:1:
../src/node_http2_core-inl.h:145:17: warning: comparison of integers of different signs: 'ssize_t' (aka 'long') and 'size_t' (aka 'unsigned long')
      [-Wsign-compare]
      if (ncopy > buf.len)
          ~~~~~ ^ ~~~~~~~
In file included from ../src/node_http2.cc:3:
In file included from ../src/node_http2.h:6:
../src/node_http2_core-inl.h:145:17: warning: comparison of integers of different signs: 'ssize_t' (aka 'long') and 'size_t' (aka 'unsigned long')
      [-Wsign-compare]
      if (ncopy > buf.len)
          ~~~~~ ^ ~~~~~~~
1 warning generated.
```

**Comment by jasnell** [comment]
@sebdeckers ... yeah, I've got a note to resolve that compiler warning before final cleanup. It's minor at this point.

I'll take a look at #163! :-)

File: test/parallel/test-http2-head-request.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+  const req = client.request({
+    [HTTP2_HEADER_METHOD]: HTTP2_METHOD_HEAD,
+    [HTTP2_HEADER_PATH]: '/'
+  });
+
+  // Because it is a HEAD request, the payload is meaningless. The
+  // option.endStream flag is set automatically making the stream
+  // non-writable.
+  req.on('error', common.mustCall((err) => assert(errCheck(err))));
```
Comment (sebdeckers) [inline test/parallel/test-http2-head-request.js]: `assert` is redundant here, can simplify to:

```js
req.on('error', common.mustCall(errCheck));
```

File: test/parallel/test-http2-head-request.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  type: Error,
+  message: 'write after end'
+});
+
+const {
+  HTTP2_HEADER_PATH,
+  HTTP2_HEADER_METHOD,
+  HTTP2_HEADER_STATUS,
+  HTTP2_METHOD_HEAD,
+} = http2.constants;
```
Comment (sebdeckers) [inline test/parallel/test-http2-head-request.js:18]: What's the purpose of using these constants instead of literal strings? 🤔

Comment (jasnell) [reply test/parallel/test-http2-head-request.js:18]: Just less error prone

**Comment by jasnell** [comment]
CI: https://ci.nodejs.org/job/node-test-pull-request/8883/

**Comment by jasnell** [comment]
CI is good. ping @mcollina @sebdeckers

File: lib/internal/http2/core.js [hunk 3]
```diff
@@ -1547,7 +1584,7 @@ class ServerHttp2Stream extends Http2Stream {
     if (state.headersSent)
       throw new errors.Error('ERR_HTTP2_HEADERS_SENT');
 
-    // Close the writable side if the endStream option is set
+    // Close the writable side of the endStream option is set
```
Comment (sebdeckers) [inline lib/internal/http2/core.js:1587]: Typo unfix?

**Review by sebdeckers — APPROVED** [review]
LGTM 🤘

@jasnell Also, how does one trigger these CI runs? Do/can they run for my PRs too? 😲

**Comment by jasnell** [comment]
The ci runs can be triggered by anyone with commit access to the main nodejs/node repo.

**Comment by jasnell** [comment]
Landed with nits fixed

**Comment by refack** [comment]
One of the CI jobs has hanged:
```
[----------] 10 tests from InspectorSocketServerTest
[ RUN      ] InspectorSocketServerTest.InspectorSessions
[       OK ] InspectorSocketServerTest.InspectorSessions (0 ms)
[ RUN      ] InspectorSocketServerTest.ServerDoesNothing
[       OK ] InspectorSocketServerTest.ServerDoesNothing (0 ms)
[ RUN      ] InspectorSocketServerTest.ServerWithoutTargets
[       OK ] InspectorSocketServerTest.ServerWithoutTargets (0 ms)
[ RUN      ] InspectorSocketServerTest.ServerCannotStart
[       OK ] InspectorSocketServerTest.ServerCannotStart (0 ms)
[ RUN      ] InspectorSocketServerTest.StoppingServerDoesNotKillConnections
[       OK ] InspectorSocketServerTest.StoppingServerDoesNotKillConnections (0 ms)
[ RUN      ] InspectorSocketServerTest.ClosingConnectionReportsDone
[       OK ] InspectorSocketServerTest.ClosingConnectionReportsDone (0 ms)
[ RUN      ] InspectorSocketServerTest.ClosingSocketReportsDone
[       OK ] InspectorSocketServerTest.ClosingSocketReportsDone (0 ms)
[ RUN      ] InspectorSocketServerTest.TerminatingSessionReportsDone
[       OK ] InspectorSocketServerTest.TerminatingSessionReportsDone (0 ms)
[ RUN      ] InspectorSocketServerTest.FailsToBindToNodejsHost
[       OK ] InspectorSocketServerTest.FailsToBindToNodejsHost (15 ms)
[ RUN      ] InspectorSocketServerTest.BindsToIpV6
```
https://ci.nodejs.org/job/node-test-binary-windows/9513/RUN_SUBSET=3,VS_VERSION=vs2015,label=win2008r2/
(I think I saw this before with `http2` code...)

**Comment by jasnell** [comment]
Very strange, especially given that the test that's hanging isn't an http2 test!

**Pull request closed without merging**